Modules/Filtering/ImageFilterBase/test/itkImageIterationPerformanceTest.cxx

```diff
@@ -0,0 +1,382 @@
+      const InputPixelType & inputPixel = inputIt.Get();
+      OutputPixelType        value(outputIt.Get());
+      for (unsigned int k = 0; k < componentsPerPixel; ++k)
+      {
+        value[k] = static_cast<typename OutputPixelType::ValueType>(inputPixel[k]);
+      }
+      outputIt.Set(value);
+
+      ++inputIt;
+      ++outputIt;
+    }
+    inputIt.NextLine();
+    outputIt.NextLine();
+  }
+}
+
+
+// Method 1b: ImageScanlineIterator approach using NumericTraits::GetLength()
+template <typename TInputImage, typename TOutputImage>
+void
+CopyScanlineIteratorNumericTraits(const TInputImage * inputPtr, TOutputImage * outputPtr)
+{
+  using InputPixelType = typename TInputImage::PixelType;
+  using OutputPixelType = typename TOutputImage::PixelType;
+  using ImageScanlineConstIterator = itk::ImageScanlineConstIterator<TInputImage>;
+  using ImageScanlineIterator = itk::ImageScanlineIterator<TOutputImage>;
+
+  const typename TOutputImage::RegionType outputRegion = outputPtr->GetRequestedRegion();
+  typename TInputImage::RegionType        inputRegion = outputRegion;
+
+  ImageScanlineConstIterator inputIt(inputPtr, inputRegion);
+  ImageScanlineIterator      outputIt(outputPtr, outputRegion);
+
+  unsigned int componentsPerPixel = itk::NumericTraits<OutputPixelType>::GetLength(outputIt.Get());
+  while (!inputIt.IsAtEnd())
+  {
+    while (!inputIt.IsAtEndOfLine())
+    {
+      const InputPixelType & inputPixel = inputIt.Get();
+
+      OutputPixelType value{ outputIt.Get() };
+      for (unsigned int k = 0; k < componentsPerPixel; ++k)
+      {
+        value[k] = static_cast<typename OutputPixelType::ValueType>(inputPixel[k]);
+      }
+      outputIt.Set(value);
+
+      ++inputIt;
+      ++outputIt;
+    }
+    inputIt.NextLine();
+    outputIt.NextLine();
+  }
+}
+
+
+// Method 2: ImageRegionRange approach
+template <typename TInputImage, typename TOutputImage>
+void
+CopyImageRegionRange(const TInputImage * inputPtr, TOutputImage * outputPtr)
+{
+  using InputPixelType = typename TInputImage::PixelType;
+  using OutputPixelType = typename TOutputImage::PixelType;
+
+  const typename TOutputImage::RegionType outputRegion = outputPtr->GetRequestedRegion();
+  typename TInputImage::RegionType        inputRegion = outputRegion;
+
+  auto inputRange = itk::ImageRegionRange<const TInputImage>(*inputPtr, inputRegion);
+  auto outputRange = itk::ImageRegionRange<TOutputImage>(*outputPtr, outputRegion);
+
+  auto       inputIt = inputRange.begin();
+  auto       outputIt = outputRange.begin();
+  const auto inputEnd = inputRange.end();
+
+  const unsigned int componentsPerPixel = inputPtr->GetNumberOfComponentsPerPixel();
+  while (inputIt != inputEnd)
+  {
+    InputPixelType  inputPixel = *inputIt;
+    OutputPixelType outputPixel(componentsPerPixel);
+    for (unsigned int k = 0; k < componentsPerPixel; ++k)
+    {
+      outputPixel[k] = static_cast<typename OutputPixelType::ValueType>(inputPixel[k]);
+    }
+    *outputIt = outputPixel;
```

Comment:
Please consider "moving" `outputPixel` into `*outputIt`. This is especially relevant when OutputPixelType is a VariableLengthVector, which has an expensive copy-constructor and a fast move-constructor.